File: samples/replicatedentity-examples/src/main/java/com/example/replicated/registermap/domain/SomeRegisterMap.java
package com.example.replicated.registermap.domain;

import com.akkaserverless.javasdk.replicatedentity.ReplicatedEntityContext;
import com.akkaserverless.javasdk.replicatedentity.ReplicatedRegisterMap;
import com.example.replicated.registermap.SomeRegisterMapApi;
import com.google.protobuf.Empty;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

public class SomeRegisterMap extends AbstractSomeRegisterMap {
  @SuppressWarnings("unused")
  private final String entityId;

  public SomeRegisterMap(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  // tag::update[]
  @Override
  public Effect<Empty> set(
      ReplicatedRegisterMap<SomeRegisterMapDomain.SomeKey, SomeRegisterMapDomain.SomeValue>
          registerMap,
      SomeRegisterMapApi.SetValue command) {
    SomeRegisterMapDomain.SomeKey key = // <1>
        SomeRegisterMapDomain.SomeKey.newBuilder().setKey(command.getKey().getKey()).build();
    SomeRegisterMapDomain.SomeValue value = // <2>
        SomeRegisterMapDomain.SomeValue.newBuilder()
            .setValue(command.getValue().getValue())
            .build();
    return effects()
        .update(registerMap.setValue(key, value)) // <3>
        .thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> remove(
      ReplicatedRegisterMap<SomeRegisterMapDomain.SomeKey, SomeRegisterMapDomain.SomeValue>
          registerMap,
      SomeRegisterMapApi.RemoveValue command) {
    SomeRegisterMapDomain.SomeKey key = // <1>
        SomeRegisterMapDomain.SomeKey.newBuilder().setKey(command.getKey().getKey()).build();
    return effects()
        .update(registerMap.remove(key)) // <3>
        .thenReply(Empty.getDefaultInstance());
  }
  // end::update[]

  // tag::get[]
  @Override
  public Effect<SomeRegisterMapApi.CurrentValue> get(
      ReplicatedRegisterMap<SomeRegisterMapDomain.SomeKey, SomeRegisterMapDomain.SomeValue>
          registerMap,
      SomeRegisterMapApi.GetValue command) {
    SomeRegisterMapDomain.SomeKey key = // <1>
        SomeRegisterMapDomain.SomeKey.newBuilder().setKey(command.getKey().getKey()).build();
    Optional<SomeRegisterMapDomain.SomeValue> maybeValue = registerMap.getValue(key); // <2>
    SomeRegisterMapApi.CurrentValue currentValue =
        SomeRegisterMapApi.CurrentValue.newBuilder()
            .setValue(
                SomeRegisterMapApi.Value.newBuilder()
                    .setValue(maybeValue.map(SomeRegisterMapDomain.SomeValue::getValue).orElse("")))
            .build();
    return effects().reply(currentValue);
  }

  @Override
  public Effect<SomeRegisterMapApi.CurrentValues> getAll(
      ReplicatedRegisterMap<SomeRegisterMapDomain.SomeKey, SomeRegisterMapDomain.SomeValue> registerMap,
      SomeRegisterMapApi.GetAllValues command) {
    List<SomeRegisterMapApi.CurrentValue> allValues =
        registerMap.keySet().stream() // <3>
            .map(
                key -> {
                  String value =
                      registerMap
                          .getValue(key)
                          .map(SomeRegisterMapDomain.SomeValue::getValue)
                          .orElse("");
                  return SomeRegisterMapApi.CurrentValue.newBuilder()
                      .setKey(SomeRegisterMapApi.Key.newBuilder().setKey(key.getKey()))
                      .setValue(SomeRegisterMapApi.Value.newBuilder().setValue(value))
                      .build();
                })
            .collect(Collectors.toList());
    SomeRegisterMapApi.CurrentValues currentValues =
        SomeRegisterMapApi.CurrentValues.newBuilder().addAllValues(allValues).build();
    return effects().reply(currentValues);
  }
  // end::get[]
}

File: samples/scala-protobuf-replicatedentity-examples/src/main/scala/com/example/replicated/registermap/domain/SomeRegisterMap.scala
package com.example.replicated.registermap.domain

import kalix.scalasdk.replicatedentity.ReplicatedEntity
import kalix.scalasdk.replicatedentity.ReplicatedEntityContext
import kalix.scalasdk.replicatedentity.ReplicatedRegisterMap
import com.example.replicated.registermap
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Kalix tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

class SomeRegisterMap(context: ReplicatedEntityContext) extends AbstractSomeRegisterMap {

  // tag::update[]
  def set(currentData: ReplicatedRegisterMap[SomeKey, SomeValue], setValue: registermap.SetValue): ReplicatedEntity.Effect[Empty] = {
    val key = SomeKey(setValue.getKey.field) // <1>
    val value = SomeValue(setValue.getValue.field) // <2>
    effects
      .update(currentData.setValue(key,value)) // <3>
      .thenReply(Empty.defaultInstance)
  }

  def remove(currentData: ReplicatedRegisterMap[SomeKey, SomeValue], removeValue: registermap.RemoveValue): ReplicatedEntity.Effect[Empty] = {
    val key = SomeKey(removeValue.getKey.field) // <1>
    effects
      .update(currentData.remove(key)) // <3>
      .thenReply(Empty.defaultInstance)
  }
  // end::update[]

  // tag::get[]
  def get(currentData: ReplicatedRegisterMap[SomeKey, SomeValue], getValue: registermap.GetValue): ReplicatedEntity.Effect[registermap.CurrentValue] = {
    val key = SomeKey(getValue.getKey.field) // <1>
    val maybeValue = currentData.get(key) // <2>
    val currentValue = registermap.CurrentValue(getValue.key, maybeValue.map(v => registermap.Value(v.someField)))
    effects.reply(currentValue)
  }

  def getAll(currentData: ReplicatedRegisterMap[SomeKey, SomeValue], getAllValues: registermap.GetAllValues): ReplicatedEntity.Effect[registermap.CurrentValues] = {

    val allData =
      currentData.keySet.map { key => // <3>
        val value = currentData.get(key).map(v => registermap.Value(v.someField))
        registermap.CurrentValue(Some(registermap.Key(key.someField)), value)
      }.toSeq

    effects.reply(registermap.CurrentValues(allData))
  }
  // end::get[]

}
